packages/@lwc/engine-core/src/framework/stylesheet.ts

```diff
@@ -22,127 +21,87 @@ export type StylesheetFactory = (
+        newShadowAttribute = newStylesheetTokens.shadowAttribute;
+
+        renderer.setAttribute(elm, newHostAttribute, '');
+    }
+
+    // Update the styling tokens present on the context object.
+    context.hostAttribute = newHostAttribute;
+    context.shadowAttribute = newShadowAttribute;
 }
 
-/**
- * Apply/Update the styling token applied to the host element.
- */
-export function applyStyleAttributes(vm: VM, hostAttribute: string, shadowAttribute: string): void {
-    const { context, elm, renderer } = vm;
+export function getStylesheetsContent(vm: VM, template: Template): string[] {
+    const { stylesheets: factories, stylesheetTokens: tokens } = template;
+    const { syntheticShadow: useSyntheticShadow } = vm.renderer;
 
-    // Remove the style attribute currently applied to the host element.
-    renderer.setAttribute(elm, hostAttribute, '');
+    const stylesheets: string[] = [];
 
-    context.hostAttribute = hostAttribute;
-    context.shadowAttribute = shadowAttribute;
-}
+    if (!isUndefined(factories) && !isUndefined(tokens)) {
+        const hostSelector = useSyntheticShadow ? `[${tokens.hostAttribute}]` : '';
+        const shadowSelector = useSyntheticShadow ? `[${tokens.shadowAttribute}]` : '';
 
-function collectStylesheets(
-    stylesheets: StylesheetFactory[],
-    hostSelector: string,
-    shadowSelector: string,
-    isNative: boolean,
-    aggregatorFn: (content: string) => void
-): void {
-    forEach.call(stylesheets, (sheet) => {
-        if (isArray(sheet)) {
-            collectStylesheets(sheet, hostSelector, shadowSelector, isNative, aggregatorFn);
-        } else {
-            aggregatorFn(sheet(hostSelector, shadowSelector, isNative));
-        }
-    });
-}
+        for (let i = 0; i < factories.length; i++) {
+            const factory = factories[i];
 
-export function evaluateCSS(
-    vm: VM,
-    stylesheets: StylesheetFactory[],
-    hostAttribute: string,
-    shadowAttribute: string
-): VNode | null {
-    if (process.env.NODE_ENV !== 'production') {
-        assert.isTrue(isArray(stylesheets), `Invalid stylesheets.`);
+            if (isArray(factory)) {
+                for (let j = 0; j < factory.length; j++) {
+                    stylesheets.push(factory[j](hostSelector, shadowSelector, !useSyntheticShadow));
+                }
+            } else {
+                stylesheets.push(factory(hostSelector, shadowSelector, !useSyntheticShadow));
+            }
+        }
     }
 
-    if (vm.renderer.syntheticShadow) {
-        const hostSelector = `[${hostAttribute}]`;
-        const shadowSelector = `[${shadowAttribute}]`;
+    return stylesheets;
+}
+
+export function createStylesheet(vm: VM, stylesheets: string[]): VNode | null {
+    const { renderer } = vm;
 
-        collectStylesheets(stylesheets, hostSelector, shadowSelector, false, (textContent) => {
-            insertGlobalStyle(textContent);
-        });
+    if (renderer.syntheticShadow) {
```

Comment:
its probably me, but: it itches to me that `insertGlobalStylesheet` is only used when `syntheticShadow` is on, and we depend on the 2 of them here, it is also another place where we do the distinction between native/syntheticShadow.

i wonder if we could use only one? ex: `getRenderedStylesheetContent` (or similar), and in that method impl, decide what to do based on the `syntheticShadow` flag.

ex:
```js
// renderer
   getRenderedStylesheetContent(stylesheets) {
        if (this.useSyntheticShadow) { // or no, it does not need to depend in synthetic shadow
             // do the global thing here
             return null;
        } else {
             return ArrayJoin.call(stylesheets, '\n');
        }
   }

// stylesheet.ts
    const shadowStyleSheetContent = renderer.getRenderedStylesheetContent(stylesheets);
    return shadowStyleSheetContent !== null ? createShadowStyleVNode(shadowStyleSheetContent) : null;
```

Reply:
I originally did the change you requested @jodarove, but I rolled it back before commit. 😆 

Here is the reason why. IMO having an `insertGlobalStyleSheet` method that is dedicated to synthetic shadow is the right approach here. I agree it is less elegant than having a single method to handle style injection both synthetic and native shadow, but the `insertGlobalStyleSheet` method name is really descriptive and will be easier to maintain in the long run. The fact that `getRenderedStyleSheetContent` does a side effect and inject the style at the document level is way less obvious.